Expand docs/examples for TCP `set_nonblocking` methods.

frewsxcv opened the pull request:
Part of https://github.com/rust-lang/rust/issues/44050.

rust-highfive commented:
r? @aturon

(rust_highfive has picked a reviewer for you, use r? to override)

sfackler commented on src/libstd/net/tcp.rs:
```diff
@@ -498,18 +498,45 @@ impl TcpStream {
+    /// This will result in `read`, `write`, `recv` and `send` operations
+    /// becoming nonblocking, i.e. immediately returning from their calls.
+    /// If the IO operation is successful, `Ok` is returned and no further
+    /// action is required. If the IO operation is pending, an error with kind
```
Saying that the operation is pending might be a bit misleading - `WouldBlock` means that the request couldn't be satisfied at that moment and you need to retry later. It's not the case that the operation's been queued up and will eventually finish.

frewsxcv replied:
thanks for mentioning this, was a little confused how it worked behind the scenes. i addressed it in https://github.com/rust-lang/rust/pull/45227/commits/6c20fe50fd4180ce89f15b2fac93cc580747d1d8 but let me know if you know a better way of wording it

kennytm commented on src/libstd/net/tcp.rs:
```diff
@@ -498,18 +498,45 @@ impl TcpStream {
+    ///     };
+    /// };
+    /// println!("number of bytes read: {}", num_bytes_read);
+    /// println!("bytes: {:?}", buf);`
```
Extra `` ` ``.

frewsxcv replied:
https://github.com/rust-lang/rust/pull/45227/commits/a18e98915dd37fd4a92623527206b24a8acb666a

tbu- commented on src/libstd/net/tcp.rs:
```diff
@@ -498,18 +498,45 @@ impl TcpStream {
+    /// [`io::ErrorKind::WouldBlock`] is returned.
+    ///
+    /// On Unix platforms, calling this method corresponds to calling `fcntl`.
+    /// On Windows calling this method corresponds to calling `ioctlsocket`.
```
It is somewhat import to specify which `fcntl` or `ioctlsocket` is used, both `fcntl` and `ioctlsocket` are some kind of "catch-all" syscall that can do wildly differing things. The `fcntl` `FIONBIO` is used on Unix platforms and the `ioctlsocket` `FIONBIO` is used on Windows.

frewsxcv replied:
addressed in https://github.com/rust-lang/rust/pull/45227/commits/c978e7317142d9cea7c88b329924438116516fdc

frewsxcv commented:
all comments have been addressed, let me know if y'all see anything else

tbu- commented:
LGTM

sfackler commented on src/libstd/net/tcp.rs:
```diff
@@ -498,18 +498,47 @@ impl TcpStream {
     ///
     /// # Examples
     ///
+    /// *Note: the usage of [busy-looping] in the following example is used
```
I feel kind of conflicted about these examples. We don't want to get too far down the epoll/IOCP rabbit hole, but it seems pretty bad for the example here to be a thing you should never do. We could maybe have some fake `wait_for_more_data()` function and add something like "typically implemented via platform-specific APIs such as epoll or IOCP" into the docs?

frewsxcv replied:
<details><summary>is this sorta what you mean?</summary>

```diff
diff --git a/src/libstd/net/tcp.rs b/src/libstd/net/tcp.rs
index 51269d57e6..946584bcf2 100644
--- a/src/libstd/net/tcp.rs
+++ b/src/libstd/net/tcp.rs
@@ -529,7 +529,10 @@ impl TcpStream {
     /// let num_bytes_read = loop {
     ///     match stream.read_to_end(&mut buf) {
     ///         Ok(b) => break b,
-    ///         Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => (),
+    ///         Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {
+    ///             wait_for_fd(); // typically implemented via platform-specific
+    ///                            // APIs such as epoll or IOCP
+    ///         }
     ///         Err(e) => panic!("encountered IO error: {}", e),
     ///     };
     /// };
@@ -839,13 +842,20 @@ impl TcpListener {
     /// for stream in listener.incoming() {
     ///     let mut stream = match stream {
     ///         Ok(s) => s,
-    ///         Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => continue,
+    ///         Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {
+    ///             wait_for_more_data(); // typically implemented via platform-specific
+    ///                                   // APIs such as epoll or IOCP
+    ///             continue;
+    ///         }
     ///         Err(e) => panic!("encountered IO error: {}", e),
     ///     };
     ///     loop {
     ///         match stream.write_all(b"foo") {
     ///             Ok(()) => break,
-    ///             Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => (),
+    ///             Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {
+    ///                 wait_for_fd(); // typically implemented via platform-specific
+    ///                                // APIs such as epoll or IOCP
+    ///             }
     ///             Err(e) => panic!("encountered IO error: {}", e),
     ///         };
     ///     }
```
<details>

sfackler replied:
Yeah, something like that seems better than nothing.

frewsxcv commented:
@sfackler latest commit should hopefully address your concerns.

i also removed the note about busy-looping now that i've added in `wait_for_fd();`, since afaik, this is no longer busy-looping

sfackler commented on src/libstd/net/tcp.rs:
```diff
@@ -780,17 +809,56 @@ impl TcpListener {
+    ///         Err(e) => panic!("encountered IO error: {}", e),
+    ///     };
+    ///     loop {
+    ///         match stream.write_all(b"foo") {
```
`write_all` isn't going to behave properly in a nonblocking world. If `fo` was successfully written but we hit a `WouldBlock` finishing the `o`, everything will be sent again the next round.

sfackler replied:
We'd also need to set the stream to nonblocking mode for the WouldBlock case to come up in the first place.

sfackler commented on src/libstd/net/tcp.rs:
```diff
@@ -780,17 +809,56 @@ impl TcpListener {
     /// listener.set_nonblocking(true).expect("Cannot set non-blocking");
+    ///
+    /// # fn wait_for_fd() { unimplemented!() }
+    /// for stream in listener.incoming() {
```
Would it be simpler to just open a socket and write to it rather than also including a listener?

frewsxcv replied:
this doc comment is for`TcpListener::set_nonblocking`, so it made sense for me to include the listener. can you expand on what you mean?

frewsxcv replied:
unless you're saying i shouldn't write to the stream. i could just read from it and print out the result. i'd also avoid the `write_all` thing from your other comment

sfackler replied:
Oh derp, sorry. I was thinking this was on `write` for some reason...

sfackler commented on src/libstd/net/tcp.rs:
```diff
@@ -498,18 +498,47 @@ impl TcpStream {
+    /// # fn wait_for_fd() { unimplemented!() }
+    /// let mut buf = vec![];
+    /// let num_bytes_read = loop {
+    ///     match stream.read_to_end(&mut buf) {
```
One last thing (sorry!) - `num_bytes_read` will only contain the number of bytes read on the final call, so we probably shouldn't save it here.

frewsxcv replied:
no reason to be sorry! i feel bad i'm dragging this on for you 😬

sfackler commented on src/libstd/net/tcp.rs:
```diff
@@ -780,17 +808,59 @@ impl TcpListener {
     ///
-    /// On Unix this corresponds to calling fcntl, and on Windows this
-    /// corresponds to calling ioctlsocket.
+    /// This will result in `read`, `write`, `recv` and `send` operations
```
`accept` here since it's a listener.

sfackler commented on src/libstd/net/tcp.rs:
```diff
@@ -780,17 +808,59 @@ impl TcpListener {
+    ///     stream.set_nonblocking(true);
+    ///     let mut buf = vec![];
+    ///     loop {
+    ///         match stream.read_to_end(&mut buf) {
```
This got added back?

frewsxcv replied:
before i was doing tcp writing, this is now doing tcp reading to avoid deadling with the `write_all` issue. i can remove reading/writing altogether and just grab the stream from the listener, which might make things simpler but not as realistic, but i'm not opposed

sfackler replied:
One option is something like `Ok(s) => handle_connection(s)`.

sfackler commented:
Looks good! Can you squash down to a single commit? r=me otherwise.

kennytm commented on src/libstd/net/tcp.rs:
```diff
@@ -780,17 +808,48 @@ impl TcpListener {
+    ///     match stream {
+    ///         Ok(s) => {
+    ///             // do something with the TcpStream
+    ///             handle_connection(s),
```
The `,` should be a `;`.

```
[01:04:21] ---- net/tcp.rs - net::tcp::TcpListener::set_nonblocking (line 826) stdout ----
[01:04:21] error: expected one of `.`, `;`, `?`, `}`, or an operator, found `,`
[01:04:21] --> net/tcp.rs:16:33
[01:04:21] |
[01:04:21] 16 | handle_connection(s),
[01:04:21] | ^ expected one of `.`, `;`, `?`, `}`, or an operator here
[01:04:21]
[01:04:21] thread 'rustc' panicked at 'couldn't compile the test', /checkout/src/librustdoc/test.rs:287:12
[01:04:21] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:04:21]
[01:04:21]
[01:04:21] failures:
[01:04:21] net/tcp.rs - net::tcp::TcpListener::set_nonblocking (line 826)
```

frewsxcv commented:
thanks for your patience and insight @sfackler ❤️ 

@bors r=sfackler rollup

bors commented:
:pushpin: Commit fe3ed20 has been approved by `sfackler`

<!-- @bors r=sfackler fe3ed20d8d72e4b2e4371da423d006021c7b6b97 -->
